primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Offchain Storage Lock
+//!
+//! A storage-based lock with a defined expiry time.
+//!
+//! The lock is using Local Storage and allows synchronizing
+//! access to critical section of your code for concurrently running Off-chain Workers.
+//! Usage of `PERSISTENT` variant of the storage persists the lock value even in case of re-orgs.
+//!
+//! A use case for the lock is to make sure that a particular section of the code is only run
+//! by one Off-chain Worker at the time. This may include performing a side-effect (i.e. an HTTP call)
+//! or alteration of single or multiple Local Storage entries.
+//!
+//! One use case would be collective updates of multiple data items
+//! or append / remove of i.e. sets, vectors which are stored in
+//! the offchain storage DB.
```

Comment:
I should note that locks do not guarantee data consistency in this case. If multiple values are mapped to multiple database keys, they all need to be updated in the same database transaction. Otherwise you may end up with only some of the values persisted in case of crash or process termination. 
So instead of locks, I'd consider exposing some kind of transaction mechanism for this. Putting each value in a separate transaction here is rather inefficient and doe not provide any consistency guarantees:
https://github.com/paritytech/substrate/blob/88248c040fd0a0ff004f448fc8c47fdf000b653e/client/db/src/offchain.rs#L62

Reply:
This sounds it that should be done in separate PR and is tangential to exclusive access IIUC :)

Reply:
Tracked in https://github.com/paritytech/substrate/issues/6139